File: worker-python/worker.py
import os
import sys
import time
import json
import uuid
import socket
import paho.mqtt.client as mqtt
import grpc
import google.generativeai as genai
from datetime import datetime

# Importar proto generado
import worker_pb2
import worker_pb2_grpc

# Configuración
WORKER_ID = f"python-worker-{socket.gethostname()}-{uuid.uuid4().hex[:8]}"
MQTT_BROKER = os.getenv('MQTT_BROKER', 'mosquitto')
MQTT_PORT = int(os.getenv('MQTT_PORT', '1883'))
WORKER_LANGUAGE = "Python"

# Estado del worker
current_status = "idle"
mqtt_client = None

def publish_log(message):
    """Publicar log en MQTT"""
    if mqtt_client and mqtt_client.is_connected():
        log_msg = {
            'timestamp': int(time.time() * 1000),
            'source': WORKER_ID,
            'message': message
        }
        mqtt_client.publish('upb/logs', json.dumps(log_msg))
        print(f"[LOG] {message}")

def register_worker():
    """Registrar worker con el master"""
    register_msg = {
        'worker_id': WORKER_ID,
        'language': WORKER_LANGUAGE,
        'status': 'idle',
        'timestamp': int(time.time() * 1000)
    }
    mqtt_client.publish('upb/workers/register', json.dumps(register_msg))
    publish_log(f"Worker {WORKER_ID} registrado")

def update_status(status):
    """Actualizar estado del worker"""
    global current_status
    current_status = status
    status_msg = {
        'worker_id': WORKER_ID,
        'status': status,
        'timestamp': int(time.time() * 1000)
    }
    mqtt_client.publish('upb/workers/status', json.dumps(status_msg))
    publish_log(f"Worker {WORKER_ID} cambió estado a: {status}")

def query_gemini(query, api_key):
    """Realizar consulta a Gemini AI"""
    try:
        genai.configure(api_key=api_key)
        model = genai.GenerativeModel('gemini-pro')
        
        publish_log(f"Worker {WORKER_ID} consultando a Gemini AI")
        response = model.generate_content(query)
        
        return response.text
    except Exception as e:
        error_msg = f"Error al consultar Gemini: {str(e)}"
        publish_log(error_msg)
        return f"Error: {error_msg}"

def simulate_long_processing():
    """Simular procesamiento largo (10 segundos como en el enunciado)"""
    sleep_time = 10
    publish_log(f"Worker {WORKER_ID} simulando procesamiento de {sleep_time}s")
    time.sleep(sleep_time)

def send_result_to_master(grpc_endpoint, result_data):
    """Enviar resultado al master via gRPC"""
    try:
        channel = grpc.insecure_channel(grpc_endpoint)
        stub = worker_pb2_grpc.WorkerCallbackStub(channel)
        
        result = worker_pb2.TaskResult(
            worker_id=result_data['worker_id'],
            session_id=result_data['session_id'],
            original_query=result_data['original_query'],
            ai_response=result_data['ai_response'],
            api_key=result_data['api_key'],
            processing_time_ms=result_data['processing_time_ms'],
            query_timestamp=result_data['query_timestamp'],
            completion_timestamp=result_data['completion_timestamp']
        )
        
        response = stub.SendResult(result)
        publish_log(f"Worker {WORKER_ID} envió resultado via gRPC: {response.message}")
        
        channel.close()
        return response.success
    except Exception as e:
        error_msg = f"Error al enviar resultado via gRPC: {str(e)}"
        publish_log(error_msg)
        print(f"[ERROR] {error_msg}", file=sys.stderr)
        return False

def process_task(task_data):
    """Procesar una tarea asignada"""
    publish_log(f"Worker {WORKER_ID} procesando tarea para sesión {task_data['session_id']}")
    update_status('busy')
    
    start_time = time.time()
    
    try:
        # 1. Consultar a Gemini
        ai_response = query_gemini(task_data['query'], task_data['api_key'])
        
        # 2. Simular procesamiento largo
        simulate_long_processing()
        
        # 3. Calcular tiempo de procesamiento
        end_time = time.time()
        processing_time_ms = int((end_time - start_time) * 1000)
        
        # 4. Preparar resultado
        result_data = {
            'worker_id': WORKER_ID,
            'session_id': task_data['session_id'],
            'original_query': task_data['query'],
            'ai_response': ai_response,
            'api_key': task_data['api_key'],
            'processing_time_ms': processing_time_ms,
            'query_timestamp': task_data['timestamp'],
            'completion_timestamp': int(time.time() * 1000)
        }
        
        # 5. Enviar resultado al master via gRPC
        success = send_result_to_master(task_data['grpc_endpoint'], result_data)
        
        if success:
            publish_log(f"Worker {WORKER_ID} completó tarea exitosamente")
        else:
            publish_log(f"Worker {WORKER_ID} completó tarea pero hubo error en callback")
        
    except Exception as e:
        error_msg = f"Error al procesar tarea: {str(e)}"
        publish_log(error_msg)
        print(f"[ERROR] {error_msg}", file=sys.stderr)
    finally:
        # Marcar como disponible
        update_status('idle')

def on_connect(client, userdata, flags, rc):
    """Callback cuando se conecta a MQTT"""
    if rc == 0:
        print(f"[MQTT] Conectado al broker")
        publish_log(f"Worker {WORKER_ID} conectado al broker MQTT")
        
        # Suscribirse al tópico de tareas para este worker
        client.subscribe(f'upb/workers/{WORKER_ID}/tasks')
        print(f"[MQTT] Suscrito a upb/workers/{WORKER_ID}/tasks")
        
        # Registrarse con el master
        register_worker()
    else:
        print(f"[MQTT] Error de conexión: {rc}", file=sys.stderr)

def on_message(client, userdata, msg):
    """Callback cuando se recibe un mensaje MQTT"""
    try:
        task_data = json.loads(msg.payload.decode())
        print(f"[TASK] Tarea recibida: {task_data['query'][:50]}...")
        
        # Procesar tarea
        process_task(task_data)
        
    except Exception as e:
        error_msg = f"Error al procesar mensaje: {str(e)}"
        print(f"[ERROR] {error_msg}", file=sys.stderr)
        publish_log(error_msg)

def main():
    """Función principal"""
    global mqtt_client
    
    print(f"[WORKER] Iniciando {WORKER_ID}")
    print(f"[WORKER] Lenguaje: {WORKER_LANGUAGE}")
    
    # Esperar a que MQTT esté listo
    print("[WORKER] Esperando 5 segundos para que MQTT esté listo...")
    time.sleep(5)
    
    # Configurar cliente MQTT
    mqtt_client = mqtt.Client(client_id=WORKER_ID)
    mqtt_client.on_connect = on_connect
    mqtt_client.on_message = on_message
    
    # Conectar a broker
    try:
        print(f"[MQTT] Conectando a {MQTT_BROKER}:{MQTT_PORT}")
        mqtt_client.connect(MQTT_BROKER, MQTT_PORT, 60)
        
        # Loop infinito
        mqtt_client.loop_forever()
        
    except KeyboardInterrupt:
        print("\n[WORKER] Deteniendo worker...")
        publish_log(f"Worker {WORKER_ID} detenido")
        mqtt_client.disconnect()
    except Exception as e:
        print(f"[ERROR] Error fatal: {str(e)}", file=sys.stderr)
        sys.exit(1)

if __name__ == '__main__':
    main()

File: worker-java/src/main/java/upb/distribuidos/Worker.java
package upb.distribuidos;

import com.google.gson.Gson;
import com.google.gson.JsonObject;
import io.grpc.ManagedChannel;
import io.grpc.ManagedChannelBuilder;
import okhttp3.*;
import org.eclipse.paho.client.mqttv3.*;

import java.io.IOException;
import java.net.InetAddress;
import java.util.UUID;
import java.util.concurrent.TimeUnit;

public class Worker {
    private static final String WORKER_LANGUAGE = "Java";
    private static String workerId;
    private static MqttClient mqttClient;
    private static String currentStatus = "idle";
    private static final Gson gson = new Gson();

    public static void main(String[] args) {
        try {
            // Generar Worker ID
            String hostname = InetAddress.getLocalHost().getHostName();
            workerId = String.format("java-worker-%s-%s", hostname, 
                UUID.randomUUID().toString().substring(0, 8));

            System.out.println("[WORKER] Iniciando " + workerId);
            System.out.println("[WORKER] Lenguaje: " + WORKER_LANGUAGE);

            // Esperar a que MQTT esté listo
            System.out.println("[WORKER] Esperando 5 segundos para que MQTT esté listo...");
            Thread.sleep(5000);

            // Configurar y conectar MQTT
            String broker = System.getenv().getOrDefault("MQTT_BROKER", "mosquitto");
            String port = System.getenv().getOrDefault("MQTT_PORT", "1883");
            String brokerUrl = String.format("tcp://%s:%s", broker, port);

            mqttClient = new MqttClient(brokerUrl, workerId);
            
            MqttConnectOptions options = new MqttConnectOptions();
            options.setCleanSession(true);

            mqttClient.setCallback(new MqttCallback() {
                @Override
                public void connectionLost(Throwable cause) {
                    System.err.println("[MQTT] Conexión perdida: " + cause.getMessage());
                }

                @Override
                public void messageArrived(String topic, MqttMessage message) {
                    handleMessage(topic, message);
                }

                @Override
                public void deliveryComplete(IMqttDeliveryToken token) {
                }
            });

            System.out.println("[MQTT] Conectando a " + brokerUrl);
            mqttClient.connect(options);

            System.out.println("[MQTT] Conectado al broker");
            publishLog("Worker " + workerId + " conectado al broker MQTT");

            // Suscribirse al tópico de tareas
            String taskTopic = "upb/workers/" + workerId + "/tasks";
            mqttClient.subscribe(taskTopic);
            System.out.println("[MQTT] Suscrito a " + taskTopic);

            // Registrarse con el master
            registerWorker();

            System.out.println("[WORKER] Worker en ejecución...");

            // Mantener vivo
            while (true) {
                Thread.sleep(1000);
            }

        } catch (Exception e) {
            System.err.println("[ERROR] Error fatal: " + e.getMessage());
            e.printStackTrace();
            System.exit(1);
        }
    }

    private static void handleMessage(String topic, MqttMessage message) {
        try {
            String payload = new String(message.getPayload());
            JsonObject taskData = gson.fromJson(payload, JsonObject.class);

            String query = taskData.get("query").getAsString();
            System.out.println("[TASK] Tarea recibida: " + 
                query.substring(0, Math.min(50, query.length())) + "...");

            processTask(taskData);

        } catch (Exception e) {
            System.err.println("[ERROR] Error al procesar mensaje: " + e.getMessage());
            e.printStackTrace();
        }
    }

    private static void registerWorker() {
        try {
            JsonObject register = new JsonObject();
            register.addProperty("worker_id", workerId);
            register.addProperty("language", WORKER_LANGUAGE);
            register.addProperty("status", "idle");
            register.addProperty("timestamp", System.currentTimeMillis());

            mqttClient.publish("upb/workers/register", 
                new MqttMessage(gson.toJson(register).getBytes()));
            
            publishLog("Worker " + workerId + " registrado");

        } catch (Exception e) {
            System.err.println("[ERROR] Error al registrar worker: " + e.getMessage());
        }
    }

    private static void updateStatus(String status) {
        try {
            currentStatus = status;
            
            JsonObject statusMsg = new JsonObject();
            statusMsg.addProperty("worker_id", workerId);
            statusMsg.addProperty("status", status);
            statusMsg.addProperty("timestamp", System.currentTimeMillis());

            mqttClient.publish("upb/workers/status", 
                new MqttMessage(gson.toJson(statusMsg).getBytes()));
            
            publishLog("Worker " + workerId + " cambió estado a: " + status);

        } catch (Exception e) {
            System.err.println("[ERROR] Error al actualizar estado: " + e.getMessage());
        }
    }

    private static void publishLog(String message) {
        try {
            JsonObject logMsg = new JsonObject();
            logMsg.addProperty("timestamp", System.currentTimeMillis());
            logMsg.addProperty("source", workerId);
            logMsg.addProperty("message", message);

            mqttClient.publish("upb/logs", 
                new MqttMessage(gson.toJson(logMsg).getBytes()));
            
            System.out.println("[LOG] " + message);

        } catch (Exception e) {
            System.err.println("[ERROR] Error al publicar log: " + e.getMessage());
        }
    }

    private static void processTask(JsonObject taskData) {
        publishLog("Worker " + workerId + " procesando tarea para sesión " + 
            taskData.get("session_id").getAsString());
        updateStatus("busy");

        long startTime = System.currentTimeMillis();

        try {
            // 1. Consultar a Gemini
            String query = taskData.get("query").getAsString();
            String apiKey = taskData.get("api_key").getAsString();
            String aiResponse = queryGemini(query, apiKey);

            // 2. Simular procesamiento largo
            simulateLongProcessing();

            // 3. Calcular tiempo de procesamiento
            long processingTime = System.currentTimeMillis() - startTime;

            // 4. Enviar resultado al master via gRPC
            String grpcEndpoint = taskData.get("grpc_endpoint").getAsString();
            sendResultToMaster(grpcEndpoint, taskData, aiResponse, processingTime);

            publishLog("Worker " + workerId + " completó tarea exitosamente");

        } catch (Exception e) {
            System.err.println("[ERROR] Error al procesar tarea: " + e.getMessage());
            e.printStackTrace();
            publishLog("Worker " + workerId + " error al procesar tarea: " + e.getMessage());
        } finally {
            updateStatus("idle");
        }
    }

    private static String queryGemini(String query, String apiKey) {
        try {
            publishLog("Worker " + workerId + " consultando a Gemini AI");

            OkHttpClient client = new OkHttpClient.Builder()
                .connectTimeout(30, TimeUnit.SECONDS)
                .readTimeout(30, TimeUnit.SECONDS)
                .build();

            String url = "https://generativelanguage.googleapis.com/v1beta/models/gemini-pro:generateContent?key=" + apiKey;

            JsonObject content = new JsonObject();
            JsonObject parts = new JsonObject();
            parts.addProperty("text", query);
            
            com.google.gson.JsonArray partsArray = new com.google.gson.JsonArray();
            partsArray.add(parts);
            
            JsonObject contentObj = new JsonObject();
            contentObj.add("parts", partsArray);
            
            com.google.gson.JsonArray contentsArray = new com.google.gson.JsonArray();
            contentsArray.add(contentObj);
            
            content.add("contents", contentsArray);

            RequestBody body = RequestBody.create(
                gson.toJson(content),
                MediaType.parse("application/json")
            );

            Request request = new Request.Builder()
                .url(url)
                .post(body)
                .build();

            try (Response response = client.newCall(request).execute()) {
                if (!response.isSuccessful()) {
                    return "Error al consultar Gemini: " + response.code();
                }

                String responseBody = response.body().string();
                JsonObject responseJson = gson.fromJson(responseBody, JsonObject.class);
                
                return responseJson
                    .getAsJsonArray("candidates").get(0)
                    .getAsJsonObject().getAsJsonObject("content")
                    .getAsJsonArray("parts").get(0)
                    .getAsJsonObject().get("text").getAsString();
            }

        } catch (Exception e) {
            String error = "Error al consultar Gemini: " + e.getMessage();
            publishLog(error);
            return error;
        }
    }

    private static void simulateLongProcessing() {
        try {
            int sleepTime = 10000; // 10 segundos
            publishLog("Worker " + workerId + " simulando procesamiento de " + sleepTime + "ms");
            Thread.sleep(sleepTime);
        } catch (InterruptedException e) {
            Thread.currentThread().interrupt();
        }
    }

    private static void sendResultToMaster(String endpoint, JsonObject taskData, 
                                          String aiResponse, long processingTime) {
        ManagedChannel channel = null;
        try {
            String[] parts = endpoint.split(":");
            String host = parts[0];
            int port = Integer.parseInt(parts[1]);

            channel = ManagedChannelBuilder
                .forAddress(host, port)
                .usePlaintext()
                .build();

            worker.WorkerCallbackGrpc.WorkerCallbackBlockingStub stub = 
                worker.WorkerCallbackGrpc.newBlockingStub(channel);

            worker.Worker.TaskResult result = worker.Worker.TaskResult.newBuilder()
                .setWorkerId(workerId)
                .setSessionId(taskData.get("session_id").getAsString())
                .setOriginalQuery(taskData.get("query").getAsString())
                .setAiResponse(aiResponse)
                .setApiKey(taskData.get("api_key").getAsString())
                .setProcessingTimeMs(processingTime)
                .setQueryTimestamp(taskData.get("timestamp").getAsLong())
                .setCompletionTimestamp(System.currentTimeMillis())
                .build();

            worker.Worker.ResultAck response = stub.sendResult(result);
            publishLog("Worker " + workerId + " envió resultado via gRPC: " + response.getMessage());

        } catch (Exception e) {
            System.err.println("[ERROR] Error al enviar resultado via gRPC: " + e.getMessage());
            e.printStackTrace();
            publishLog("Worker " + workerId + " error al enviar resultado: " + e.getMessage());
        } finally {
            if (channel != null) {
                try {
                    channel.shutdown().awaitTermination(5, TimeUnit.SECONDS);
                } catch (InterruptedException e) {
                    Thread.currentThread().interrupt();
                }
            }
        }
    }
}
